### PR TITLE
Make satellite resend context when it suspects server cannot remember gpu

### LINKED ISSUE #134: Satellite resend context
If a satellite attempts to send data but gets an error because the database is unaware of it, it should resend the context

### PR DESCRIPTION
**Pull request by YRMYJASKA**
Closes #134

### CODE REVIEW

**Comment by YRMYJASKA** [comment]
Closes #134

File: internal/database/postgres.go [hunk 1]
```diff
@@ -176,6 +176,11 @@ func (conn postgresConn) AppendDataPoint(sample uplink.GPUStatSample) error {
 		sample.GraphicsClock, sample.MaxGraphicsClock,
 		sample.MemoryClock, sample.MaxMemoryClock)
 
+	// TODO: hacky, untested. Might not work
```
Comment (danielg0) [inline internal/database/postgres.go:179]: This is fine, because I'm about to change a fair bit of postgres.go anyway

**Review by danielg0 — APPROVED** [review]
LGTM

This does treat any 400 response as a missing GPU error, but that's fine because that's the only time we currently return a 400. If we wanted to return 400 codes for other reasons (invalid json, etc.), it would be neat to see if we could encode the Go error in the response body, then have one set of errors shared between the groundstation and satellites (so, we can use `errors.Is()`)

File: internal/femto/femto.go [hunk 2]
```diff
@@ -75,10 +75,10 @@ func Ok[T any](content T) (*Response[T], error) {
 	return &Response[T]{Status: http.StatusOK, Body: content}, nil
 }
 
-func generateErrorLogger(l *slog.Logger, w http.ResponseWriter) func(ctx string, status int, e error) {
+func generateErrorLogger(l *slog.Logger, w *http.ResponseWriter) func(ctx string, status int, e error) {
```
Comment (aDotInTheVoid) [inline internal/femto/femto.go:78]: NIT: Whats the pupose of this change? [ResponceWriter](https://pkg.go.dev/net/http#ResponseWriter) is already an interface, so should be passed by value, unless you have a specific need

File: internal/database/inmemory.go [hunk 3]
```diff
@@ -42,7 +37,7 @@ func (m *inMemory) AppendDataPoint(sample uplink.GPUStatSample) error {
 	defer m.mu.Unlock()
 
 	if info, pres := m.infos[sample.Uuid]; !pres {
-		return fmt.Errorf("%w: %s", ErrGpuNotPresent, sample.Uuid)
+		return ErrGpuNotPresent
```
Comment (aDotInTheVoid) [inline internal/database/inmemory.go:40]: Why make this change?

File: internal/groundstation/gpustats.go [hunk 4]
```diff
@@ -27,7 +28,9 @@ func (gs *groundstation) gpustats(data uplink.GpuStatsUpload, req *http.Request,
 
 	if len(data.Stats) > 0 {
 		err := gs.handleGPUStatSamples(data.Hostname, data.Stats)
-		if err != nil {
+		if err == database.ErrGpuNotPresent {
```
Comment (aDotInTheVoid) [inline internal/groundstation/gpustats.go:31]: This should use `errors.Is`, so it works on wrapped errors.

File: cmd/satellite/main.go [hunk 5]
```diff
@@ -194,14 +229,24 @@ func (s *satellite) sendGPUStatusWithSource(gpuhandler gpustats.GPUDataSource) e
...
 		uplink.GpuStatsUpload{Hostname: s.hostname, Stats: stats},
 	)
+
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode == http.StatusBadRequest {
+		// Could not send status, suspected cause is server does not have context.
+		return errSuspectedServerMissingInfo
```
Comment (aDotInTheVoid) [inline cmd/satellite/main.go:248]: NIT: Needs a todo that this assumes only 1 error.

**Review by aDotInTheVoid — COMMENTED** [review]
Gah!, I thaught I'd reviewed this, but github apparenly kept my comments in the pending state?

**Pull request merged**